Efficient bulk tree creation.

jmcarp opened the pull request:
Allow bulk-creating a new tree or subtree while setting tree properties
in memory to minimize database queries.

[Resolves #442]

@craigds @erez-o: let me know if the proposed interface makes sense--I'll be happy to revise if you have suggestions.

cc @cmc333333

jmcarp commented:
Ping--let me know if you're interested in this feature. Feedback welcome!

matthiask commented:
I like it. The code looks fine, the tests sufficient.

craigds commented:
This sounds great :) Though I have a few questions:

 * should `bulk_load` be on the manager instead of the model class?
 * The items in the tests have `id` values already. Do the nodes have to have known IDs?
 * The nodes in the tests are all related to each other. Could I use this to create completely unrelated nodes, sprinkled arbitrarily across my forest?
 * is there a way we can use existing django interfaces for this (`Manager.bulk_create`, specifically) instead of a new method with new semantics? 

Re that last point, do you think there's a way we could make this interface work?

```
a = MyNode(name='a')
b = MyNode(name='b', parent=a)
c  = MyNode(name='c')
MyNode.objects.bulk_create([a, b, c])
```

Thanks for the effort you've put into this :)

jmcarp commented:
> should bulk_load be on the manager instead of the model class?
Now that you mention it, that sounds like the right place for it. Will move.

> The items in the tests have id values already. Do the nodes have to have known IDs?
The nodes don't have to have primary keys, but if a user wants to use `bulk_create` with foreign keys, primary keys should be pre-populated before save. If we need to use auto-incrementing keys, we need to run multiple `INSERT` queries, so `bulk_create` becomes less useful.

> Could I use this to create completely unrelated nodes, sprinkled arbitrarily across my forest?
My intention was to speed up the specific use case of inserting a tree of related nodes. For inserting a bunch of unrelated nodes, I think turning off MPTT updates, running `bulk_create`, and rebuilding is probably the best option.

> is there a way we can use existing django interfaces for this (Manager.bulk_create, specifically)
Not sure--I don't think overriding `bulk_create` is right, since `bulk_load` doesn't actually run an `INSERT`, and since I think we still want to expose the original `bulk_create` method, as in the use case above when we're inserting unrelated nodes.

Basically, the proposed interface works because it's constrained. I haven't figured out how to write an generic and efficient `bulk_create` that's better than `disable_mptt_updates` and `rebuild`.

matthiask commented:
@jmcarp Are you still interested resp. do you still have the time to work on this?

jmcarp commented:
@matthiask: sorry, this fell off my radar, but I'm definitely still interested in getting this in good shape. I just moved the `bulk_load` method from `MPTTModel` to `TreeManager` as suggested, and fixed the merge conflict. Let me know if this needs more revisions.

eressler commented:
Any update on when this will make it in? Even with disable_mptt_updates() importing very large trees takes hours (nine hours and counting on my current import)

jmcarp commented:
Bump. I'm happy to revise if there's anything missing.

craigds commented on mptt/managers.py:
```diff
@@ -672,6 +672,79 @@ def partial_rebuild(self, tree_id):
         self._rebuild_helper(pks[0], 1, tree_id)
 
+    @delegate_manager
+    def bulk_load(self, data, target=None, position='last-child'):
```
From a cursory glance this method doesn't appear to support trees that use `order_insertion_by`, since it always appends nodes to the next available `tree_id`. Is that correct?

I imagine doing that would be much trickier. At a minimum I think this method should throw an error if you try to call it when using `order_insertion_by`

craigds replied:
Not convinced about the name - I think it's too close to `bulk_create` and it's not obvious that this is a special tree-related thing. If I was tab-completing in a shell I'd be unsure which one to pick.

Maybe something like:
* `build_tree_nodes`
* `load_tree_nodes`
* `tree_bulk_load`

murdav commented:
Dear All, 
What is the status of this great/lovely/excellent bulk_load?
@craigds when do you program to merge it?

Thanks!

cirotix commented:
Hi @jmcarp @craigds,

I am definitively interested to get this merged. As I need to be able to `bulk_create` efficiently for my project, I will fork and apply the patch from @jmcarp on my branch. 

As this PR seems stalled, I can do  the changes asked by @craigds and open a new PR

cirotix commented on mptt/managers.py:
```diff
@@ -672,6 +672,79 @@ def partial_rebuild(self, tree_id):
+
+        stack = []
+
+        def treeify(data, cursor=1, level=0):
```
Actually the `treeify` function uses the `stack` and `opts` variables from the enclosing scope so it should stay as it.

Tyler-Churchill commented:
Just discovered this project - looks awesome! I think this feature is a must, Any status on when this will be looked at/considered again for merge?

matthiask commented:
Closed in favour of https://github.com/django-mptt/django-mptt/pull/575
